drivers/adc/adc_ads131m02.c

```diff
@@ -0,0 +1,738 @@
+#define ADS131M02_OSR_1024_MASK		BIT(3) | BIT(2)
+#define ADS131M02_OSR_2048_MASK		BIT(4)
+#define ADS131M02_OSR_4096_MASK		BIT(4) | BIT(2)
+#define ADS131M02_OSR_8192_MASK		BIT(4) | BIT(3)
+#define ADS131M02_OSR_16384_MASK	BIT(4) | BIT(3) | BIT(2)
+#define ADS131M02_GC_MODE_MASK		BIT(8)
+#define ADS131M02_GC_DELAY_MASK		GENMASK(12, 9)
+
+#define ADS131M02_DISABLE_ADC		0x000E
+#define ADS131M02_RESET_DELAY		100
+
+enum {
+	ADS131M02_GAIN_1 = 0,
+	ADS131M02_GAIN_2 = 1,
+	ADS131M02_GAIN_4 = 2,
+	ADS131M02_GAIN_8 = 3,
+	ADS131M02_GAIN_16 = 4,
+	ADS131M02_GAIN_32 = 5,
+	ADS131M02_GAIN_64 = 6,
+	ADS131M02_GAIN_128 = 7,
+};
+
+#define ADS131M02_GET_GAIN(channel_id, gain)	\
+		FIELD_PREP(channel_id == 0 ? ADS131M02_GAIN0_MASK : \
+			   ADS131M02_GAIN1_MASK, gain)
+
+enum ads131m02_data_rate {
+	/* SPS */
+	ADS131M02_DR_250,
+	ADS131M02_DR_500,
+	ADS131M02_DR_1k,
+	ADS131M02_DR_2k,
+	ADS131M02_DR_4k,
+	ADS131M02_DR_8k,
+	ADS131M02_DR_16k,
+	ADS131M02_DR_32k,
+};
+
+struct ads131m02_config {
+	const struct spi_dt_spec spi;
+	const struct gpio_dt_spec gpio_drdy;
+};
+
+struct ads131m02_data {
+	struct adc_context ctx;
+	struct k_sem acq_sem;
+	struct k_sem drdy_sem;
+	struct gpio_callback callback_drdy;
+	int32_t *buffer;
+	int32_t *buffer_ptr;
+};
+
+static inline int ads131m02_check_device(const struct spi_dt_spec *dev)
+{
+	return spi_is_ready_dt(dev);
+}
+
+static inline int ads131m02_transceive(const struct device *dev,
+				       uint8_t *send_buf, size_t send_buf_len,
+				       uint8_t *recv_buf, size_t recv_buf_len)
+{
+	int ret = 0;
+	const struct ads131m02_config *cfg = dev->config;
+
+	struct spi_buf tx_buf = {
+		.buf = send_buf,
+		.len = send_buf_len,
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	struct spi_buf rx_buf = {
+		.buf = recv_buf,
+		.len = recv_buf_len,
+	};
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1,
+	};
+
+	ret = spi_transceive_dt(&cfg->spi, &tx, NULL);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return spi_read_dt(&cfg->spi, &rx);
+}
+
+static int ads131m02_reg_read(const struct device *dev, uint16_t addr,
+			      uint8_t *read_buf, size_t read_buf_len)
+{
+	uint16_t temp = 0;
+	uint8_t tx_buf[3] = {0};
+
+	temp = (uint16_t)(ADS131M02_RREG_CMD | (addr << 7) | 0);
+	sys_put_be16(temp, tx_buf);
+
+	return ads131m02_transceive(dev, tx_buf, sizeof(tx_buf),
+				    read_buf, read_buf_len);
+}
+
+static int ads131m02_reg_write(const struct device *dev, uint16_t addr,
+			       uint16_t write_data)
+{
+	uint16_t temp = 0;
+	uint8_t tx_buf[6] = {0};
+	uint8_t rx_buf[3] = {0};
+
+	temp = (uint16_t)(ADS131M02_WREG_CMD | (addr << 7) | 0);
+	sys_put_be16(temp, tx_buf);
+	sys_put_be16(write_data, &tx_buf[3]);
+
+	return ads131m02_transceive(dev, tx_buf, sizeof(tx_buf),
+				    rx_buf, sizeof(rx_buf));
+
+}
+
+static inline int ads131m02_configure_gain(const struct device *dev,
+					   const struct adc_channel_cfg *channel_cfg)
+{
+	uint16_t gain_cfg = 0;
```

Comment:
drop the initializer, this is actually a good case where not initializing this makes the compiler warn you if you mess up the switch

Reply:
done